nailgun/entity_mixins.py

```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
 
         """
         return self.read(attrs=self.update_json(fields))
+
+
+class EntitySearchMixin(object):
+    """This mixin provides the ability to search for or get a list of entities.
+
+    The methods provided by this class work together. The call tree looks like
+    this::
+
+        search
+        ├── search_json
+        │   └── search_raw
+        │       └── search_payload
+        ├── search_normalize
+        └── search_filter
+
+    In short, here is what the methods do:
+
+    :meth:`search_payload`
+        Assemble a search query that can be encoded and sent to the server.
+    :meth:`search_raw`
+        Make an HTTP GET request to the server, including the payload.
+    :meth:`search_json`
+        Check the server's response for errors and decode the response.
+    :meth:`search_normalize`
+        Normalize search results so they can be used to create new entities.
+    :meth:`search`
+        Create one or more :class:`nailgun.entity_mixins.Entity` objects
+        representing the found entities and populate their fields.
+    :meth:`search_filter`
+        Read all ``entities`` and locally filter them.
+
+    See the individual methods for more detailed information.
+
+    """
+
+    def search_payload(self, fields=None, query=None):
+        """Create a search query.
+
+        Do the following:
+
+        1. Generate a search query. If ``fields`` is specified, all named
+           fields are included in the query. Otherwise, all populated fields
+           are included.
+        2. Merge ``query`` in to the generated search query.
+        3. Return the result.
+
+        The rules for generating a search query are simple, and can be
+        illustrated by example::
+
+            >>> SomeEntity(
```

Comment:
I'm confused by the example here. The docstring mentions that this is an example for generating a search query... should this have used `search_payload` instead?